File: crates/meridian-realtime/src/sync/vector_clock.rs
//! Vector clock implementation for causality tracking

use std::collections::HashMap;
use std::cmp::Ordering;

use serde::{Deserialize, Serialize};

/// Vector clock for tracking causality in distributed systems
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct VectorClock {
    /// Clock values per node
    clocks: HashMap<String, u64>,

    /// This node's ID
    node_id: String,
}

impl VectorClock {
    /// Create new vector clock
    pub fn new(node_id: String) -> Self {
        let mut clocks = HashMap::new();
        clocks.insert(node_id.clone(), 0);

        Self { clocks, node_id }
    }

    /// Get clock value for a node
    pub fn get(&self, node_id: &str) -> Option<u64> {
        self.clocks.get(node_id).copied()
    }

    /// Increment this node's clock
    pub fn increment(&mut self) {
        *self.clocks.entry(self.node_id.clone()).or_insert(0) += 1;
    }

    /// Update clock for a specific node
    pub fn update(&mut self, node_id: String, value: u64) {
        self.clocks.insert(node_id, value);
    }

    /// Merge with another vector clock (take max of each component)
    pub fn merge(&mut self, other: &VectorClock) {
        for (node_id, &other_value) in &other.clocks {
            self.clocks
                .entry(node_id.clone())
                .and_modify(|our_value| {
                    *our_value = (*our_value).max(other_value);
                })
                .or_insert(other_value);
        }
    }

    /// Check if this clock happened before another
    pub fn happens_before(&self, other: &VectorClock) -> bool {
        let mut strictly_less = false;

        // Get all node IDs from both clocks
        let mut all_nodes: std::collections::HashSet<_> = self.clocks.keys().collect();
        all_nodes.extend(other.clocks.keys());

        for node_id in all_nodes {
            let our_value = self.clocks.get(node_id).copied().unwrap_or(0);
            let other_value = other.clocks.get(node_id).copied().unwrap_or(0);

            if our_value > other_value {
                return false; // Not happened before
            } else if our_value < other_value {
                strictly_less = true;
            }
        }

        strictly_less
    }

    /// Check if this clock is concurrent with another
    pub fn is_concurrent(&self, other: &VectorClock) -> bool {
        !self.happens_before(other) && !other.happens_before(self) && self != other
    }

    /// Compare with another vector clock
    pub fn compare(&self, other: &VectorClock) -> ClockOrdering {
        if self == other {
            ClockOrdering::Equal
        } else if self.happens_before(other) {
            ClockOrdering::Before
        } else if other.happens_before(self) {
            ClockOrdering::After
        } else {
            ClockOrdering::Concurrent
        }
    }

    /// Get all node IDs
    pub fn node_ids(&self) -> impl Iterator<Item = &String> {
        self.clocks.keys()
    }

    /// Get this node's ID
    pub fn node_id(&self) -> &str {
        &self.node_id
    }

    /// Get current value for this node
    pub fn value(&self) -> u64 {
        self.clocks.get(&self.node_id).copied().unwrap_or(0)
    }

    /// Create a copy with incremented value
    pub fn incremented(&self) -> Self {
        let mut new_clock = self.clone();
        new_clock.increment();
        new_clock
    }

    /// Reset clock
    pub fn reset(&mut self) {
        self.clocks.clear();
        self.clocks.insert(self.node_id.clone(), 0);
    }

    /// Get total events across all nodes
    pub fn total_events(&self) -> u64 {
        self.clocks.values().sum()
    }
}

/// Ordering between vector clocks
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ClockOrdering {
    /// Clocks are equal
    Equal,
    /// This clock is before the other
    Before,
    /// This clock is after the other
    After,
    /// Clocks are concurrent
    Concurrent,
}

impl PartialOrd for VectorClock {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        match self.compare(other) {
            ClockOrdering::Equal => Some(Ordering::Equal),
            ClockOrdering::Before => Some(Ordering::Less),
            ClockOrdering::After => Some(Ordering::Greater),
            ClockOrdering::Concurrent => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_vector_clock_creation() {
        let clock = VectorClock::new("node1".to_string());
        assert_eq!(clock.get("node1"), Some(0));
        assert_eq!(clock.value(), 0);
    }

    #[test]
    fn test_vector_clock_increment() {
        let mut clock = VectorClock::new("node1".to_string());
        clock.increment();
        assert_eq!(clock.get("node1"), Some(1));
        clock.increment();
        assert_eq!(clock.get("node1"), Some(2));
    }

    #[test]
    fn test_vector_clock_merge() {
        let mut clock1 = VectorClock::new("node1".to_string());
        clock1.increment();
        clock1.increment();

        let mut clock2 = VectorClock::new("node2".to_string());
        clock2.increment();

        clock1.merge(&clock2);

        assert_eq!(clock1.get("node1"), Some(2));
        assert_eq!(clock1.get("node2"), Some(1));
    }

    #[test]
    fn test_happens_before() {
        let mut clock1 = VectorClock::new("node1".to_string());
        clock1.increment();

        let mut clock2 = clock1.clone();
        clock2.increment();

        assert!(clock1.happens_before(&clock2));
        assert!(!clock2.happens_before(&clock1));
    }

    #[test]
    fn test_concurrent() {
        let mut clock1 = VectorClock::new("node1".to_string());
        clock1.increment();

        let mut clock2 = VectorClock::new("node2".to_string());
        clock2.increment();

        assert!(clock1.is_concurrent(&clock2));
        assert!(clock2.is_concurrent(&clock1));
    }

    #[test]
    fn test_compare() {
        let mut clock1 = VectorClock::new("node1".to_string());
        clock1.increment();

        let clock2 = clock1.clone();
        assert_eq!(clock1.compare(&clock2), ClockOrdering::Equal);

        let mut clock3 = clock1.clone();
        clock3.increment();
        assert_eq!(clock1.compare(&clock3), ClockOrdering::Before);
        assert_eq!(clock3.compare(&clock1), ClockOrdering::After);

        let mut clock4 = VectorClock::new("node2".to_string());
        clock4.increment();
        assert_eq!(clock1.compare(&clock4), ClockOrdering::Concurrent);
    }

    #[test]
    fn test_total_events() {
        let mut clock = VectorClock::new("node1".to_string());
        clock.increment();
        clock.increment();

        let mut clock2 = VectorClock::new("node2".to_string());
        clock2.increment();

        clock.merge(&clock2);
        assert_eq!(clock.total_events(), 3);
    }
}

File: crates/enterprise-collaboration/ts/src/crdt/VectorClock.ts
/**
 * Vector Clock Implementation for Distributed Systems
 * Provides causality tracking and conflict detection
 */

export type VectorClockMap = Record<string, number>;

export enum ClockComparison {
  BEFORE = 'before',
  AFTER = 'after',
  CONCURRENT = 'concurrent',
  EQUAL = 'equal',
}

export class VectorClock {
  private clock: VectorClockMap;

  constructor(initial: VectorClockMap = {}) {
    this.clock = { ...initial };
  }

  /**
   * Increment the clock for a specific node
   */
  increment(nodeId: string): void {
    this.clock[nodeId] = (this.clock[nodeId] || 0) + 1;
  }

  /**
   * Get the current value for a node
   */
  get(nodeId: string): number {
    return this.clock[nodeId] || 0;
  }

  /**
   * Set the value for a specific node
   */
  set(nodeId: string, value: number): void {
    this.clock[nodeId] = value;
  }

  /**
   * Merge with another vector clock (take maximum of each component)
   */
  merge(other: VectorClock): VectorClock {
    const merged: VectorClockMap = { ...this.clock };

    for (const [nodeId, value] of Object.entries(other.toObject())) {
      merged[nodeId] = Math.max(merged[nodeId] || 0, value);
    }

    return new VectorClock(merged);
  }

  /**
   * Compare this clock with another to determine causal relationship
   */
  compare(other: VectorClock): ClockComparison {
    const thisClock = this.clock;
    const otherClock = other.toObject();

    const allNodes = new Set([
      ...Object.keys(thisClock),
      ...Object.keys(otherClock),
    ]);

    let hasGreater = false;
    let hasLess = false;

    for (const nodeId of allNodes) {
      const thisValue = thisClock[nodeId] || 0;
      const otherValue = otherClock[nodeId] || 0;

      if (thisValue > otherValue) {
        hasGreater = true;
      } else if (thisValue < otherValue) {
        hasLess = true;
      }
    }

    if (!hasGreater && !hasLess) {
      return ClockComparison.EQUAL;
    }
    if (hasGreater && !hasLess) {
      return ClockComparison.AFTER;
    }
    if (hasLess && !hasGreater) {
      return ClockComparison.BEFORE;
    }
    return ClockComparison.CONCURRENT;
  }

  /**
   * Check if this clock happens before another
   */
  happensBefore(other: VectorClock): boolean {
    return this.compare(other) === ClockComparison.BEFORE;
  }

  /**
   * Check if this clock happens after another
   */
  happensAfter(other: VectorClock): boolean {
    return this.compare(other) === ClockComparison.AFTER;
  }

  /**
   * Check if this clock is concurrent with another
   */
  isConcurrent(other: VectorClock): boolean {
    return this.compare(other) === ClockComparison.CONCURRENT;
  }

  /**
   * Check if this clock is equal to another
   */
  isEqual(other: VectorClock): boolean {
    return this.compare(other) === ClockComparison.EQUAL;
  }

  /**
   * Clone this vector clock
   */
  clone(): VectorClock {
    return new VectorClock({ ...this.clock });
  }

  /**
   * Convert to plain object
   */
  toObject(): VectorClockMap {
    return { ...this.clock };
  }

  /**
   * Convert to JSON string
   */
  toJSON(): string {
    return JSON.stringify(this.clock);
  }

  /**
   * Create from JSON string
   */
  static fromJSON(json: string): VectorClock {
    return new VectorClock(JSON.parse(json));
  }

  /**
   * Get all node IDs in the clock
   */
  getNodeIds(): string[] {
    return Object.keys(this.clock);
  }

  /**
   * Get the sum of all clock values
   */
  getSum(): number {
    return Object.values(this.clock).reduce((sum, val) => sum + val, 0);
  }

  /**
   * Reset the clock
   */
  reset(): void {
    this.clock = {};
  }

  /**
   * Create a new vector clock
   */
  static create(nodeId?: string): VectorClock {
    const clock = new VectorClock();
    if (nodeId) {
      clock.increment(nodeId);
    }
    return clock;
  }
}
